packages/backend-core/src/sql/sql.ts

```diff
@@ -813,30 +799,66 @@ class InternalBuilder {
     return query
   }
 
+  isSqs(t?: Table): boolean {
+    const table = t || this.table
+    return (
+      table.sourceType === TableSourceType.INTERNAL ||
+      table.sourceId === INTERNAL_TABLE_SOURCE_ID
+    )
+  }
+
+  getTableName(t?: Table | string): string {
+    let table: Table
+    if (typeof t === "string") {
+      if (this.query.table?.name === t) {
+        table = this.query.table
+      } else if (this.query.meta.table?.name === t) {
+        table = this.query.meta.table
+      } else if (!this.query.meta.tables?.[t]) {
+        // This can legitimately happen in custom queries, where the user is
+        // querying against a table that may not have been imported into
+        // Budibase.
+        return t
+      } else {
+        table = this.query.meta.tables[t]
+      }
+    } else if (t) {
+      table = t
+    } else {
+      table = this.table
+    }
+
+    let name = table.name
+    if (this.isSqs(table) && table._id) {
+      // SQS uses the table ID rather than the table name
+      name = table._id
+    }
+    const aliases = this.query.tableAliases || {}
+    return aliases[name] ? aliases[name] : name
+  }
+
   addDistinctCount(query: Knex.QueryBuilder): Knex.QueryBuilder {
-    const primary = this.table.primary
-    const aliases = this.query.tableAliases
-    const aliased =
-      this.table.name && aliases?.[this.table.name]
-        ? aliases[this.table.name]
-        : this.table.name
-    if (!primary) {
+    if (!this.table.primary) {
       throw new Error("SQL counting requires primary key to be supplied")
     }
-    return query.countDistinct(`${aliased}.${primary[0]} as total`)
+    return query.countDistinct(
+      `${this.getTableName()}.${this.table.primary[0]} as total`
```

Comment:
Can we have issues here with tables containing a `total` column? Should we prefix them to avoid conflicts?

Reply:
Good question. I'll change it to be safe.